### ISSUE TITLE
Static Analysis Glitch

### ISSUE THREAD

**Issue by athleticus** [CSSE1001/MyPyTutor#163]
Static analysis incorrectly detects the following as erroneous (due to erroneous code after return):

>In MyPyTutor for CSSE1001: Functionality Programming the follow code is entered:

>```
def add_functions(f,g):
    """ Returns a function that accepts a single argument and returns f(x) + g(x)
    """
    return lambda x: f(x) + g(x)
    add = lambda f,g:f+g
    ​return add
```
>
>the following error occurs:
>
>Your lambda should only take in a single arg

**Comment by jgat** [comment]
So obviously the static analysis is only counting the number of arguments in the last lambda expression. It's probably too hard in the analysis to try and track which lambda function actually gets returned (it might even be impossible).

Probably the best solution is to add an extra check to the analysis to first give an error if there is more than one lambda function defined. It would still cause this answer to be marked wrong even when the code behaves correctly, but the error message would be more accurate. (I'm mostly in favour of this, since the student should be able to recognise and remove the redundant code, especially this late in the semester.)

**Comment by pjritee** [comment]
I agree - and maybe the question should perhaps say that a single lambda terms is required

**Comment by sapi** [comment]
It actually wouldn't be too hard to track which lambda is returned (if the
lambda is returned directly, it is a property of the return node; else, it
would/should/could be in .assigned_value) but I agree that it's not worth
the effort.

I reckon just error out if there's more than one lambda.  (Even a generic
error on any question where there's code after the return statement could
be useful.)

I would prefer not to say that a single lambda term is required in the
question; better to have students work that out themselves (with the
analyser prodding them as required).

**Comment by pjritee** [comment]
OK - fair point - making it an error to have code after return would be fine - what about

if some_test:
    return lambda x,y: ...
else:
    return lambda x: ...

**Comment by jgat** [comment]
My [proposed fix](https://github.com/CSSE1001/MyPyTutor/pull/165) is to count every lambda expression in the whole function, regardless of return values. So that example should count as two, and give the error "You only need to define one lambda function".
